File: ruby/base_item_updater.rb
class BaseItemUpdater
  def initialize(item)
    @item = item
  end

  def update_properties
    update_quality
    update_sell_in
  end

  private

  def safely_decrease_quality(value)
    value.times { @item.quality -= 1 if can_lower_quality_value? }
  end

  def safely_increase_quality(value)
    value.times { @item.quality += 1 if can_increase_quality_value? }
  end

  def drop_quality
    @item.quality = 0
  end

  def can_increase_quality_value?
    @item.quality < 50
  end

  def can_lower_quality_value?
    @item.quality > 0
  end

  def update_sell_in
    @item.sell_in -= 1
  end

  def update_quality; end
end

File: js/src/base_item_updater.js
export default class BaseItemUpdater {
  constructor(item) {
    this.item = item;
  }

  canDecreaseQualityValue() {
    return this.item.quality > 0;
  }

  canIncreaseQualityValue() {
    return this.item.quality < 50;
  }

  safelyDecreaseQuality(value) {
    for (let i = 0; i < value; i++) {
      if (this.canDecreaseQualityValue()) {
        this.item.quality -= 1;
      }
    }
  }

  safelyIncreaseQuality(value) {
    for (let i = 0; i < value; i++) {
      if (this.canIncreaseQualityValue()) {
        this.item.quality += 1;
      }
    }
  }

  dropQuality() {
    this.item.quality = 0;
  }

  updateQuality() {}

  updateSellIn() {
    this.item.sellIn -= 1;
  }

  updateProperties() {
    this.updateQuality();
    this.updateSellIn();
  }
}
